File: lectures/binary-search/sqrt.py
# Sqrt(x), https://leetcode.com/explore/learn/card/binary-search/125/template-i/950/
# Find the last M such that M*M <= X.
class Solution:
    def mySqrt(self, x: int) -> int:
        L, R = 0, x
        ans = -1 # this -1 shouldn't be returned because we will find some M such that M*M <= x, e.g. M=0 is fine
        while L <= R:
            mid = L + (R - L) // 2
            if mid * mid <= x:
                ans = mid
                L = mid + 1 # look for the last (rightmost) element satisfying the condition
            else:
                R = mid - 1
        return ans
            

File: lectures/binary-search/sqrt.cpp
// Sqrt(x), https://leetcode.com/explore/learn/card/binary-search/125/template-i/950/
/*
int f() {
	int left = 0, right = n - 1;
    int ans = -1;
	while(left <= right) {
		int mid = left + (right - left) / 2;
		if(property(mid)) {
			ans = mid;
            right = mid - 1; // if we're looking for the first (leftmost element)
		}
		else {
			left = mid + 1;
		}
	}
	return ans;
}
*/

// Find the last number M that M*M <= x.

class Solution {
public:
    int mySqrt(int x) {
       int left = 0, right = x;
        int ans = -1;
        while(left <= right) {
            int mid = left + (right - left) / 2;
            if((long long) mid * mid <= x) {
                ans = mid;
                left = mid + 1; // we're looking for the last element satisfying the condition
            }
            else {
                right = mid - 1;
            }
        }
        return ans;
    }
};
